src/OpenTelemetry.Extensions/Trace/AutoFlushActivityProcessor.cs

```diff
@@ -0,0 +1,108 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// Activity processor that flushes its containing <see cref="TracerProvider"/> if an ended
+    /// activity matches a predicate.
+    /// </summary>
+    /// <remarks>
+    /// Add this processor *after* exporter related span processors.
+    /// </remarks>
+    public sealed class AutoFlushActivityProcessor : BaseProcessor<Activity>
+    {
+        private readonly Predicate<Activity> predicate;
+        private readonly int timeoutMilliseconds;
+
+        private TracerProvider tracerProvider;
+        private bool canForceFlush = true;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoFlushActivityProcessor"/> class.
+        /// </summary>
+        /// <param name="predicate">Predicate that should return <c>true</c> to initiate a flush.</param>
+        /// <param name="timeoutMilliseconds">Timeout (in milliseconds) to use for flushing.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the <c>timeoutMilliseconds</c> is smaller than -1.
+        /// </exception>
+        /// <remarks>
+        /// It's assumed that the predicate is defined as a lambda expression which is executed quite fast and
+        /// doesn't contain more complex code. The predicate must not create new Activity instances,
+        /// otherwise the behavior is undefined. Any exception thrown by the predicate will be swallowed and logged.
+        /// </remarks>
+        public AutoFlushActivityProcessor(Predicate<Activity> predicate, int timeoutMilliseconds)
+        {
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            if (timeoutMilliseconds < Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds));
+            }
+
+            this.timeoutMilliseconds = timeoutMilliseconds;
+        }
+
+        /// <inheritdoc/>
+        public override void OnEnd(Activity data)
+        {
+            if (!this.canForceFlush)
+            {
+                return;
+            }
+
+            if (this.tracerProvider == null && this.ParentProvider != null)
+            {
+                this.tracerProvider = this.ParentProvider as TracerProvider;
+                this.canForceFlush = this.tracerProvider != null;
+                if (!this.canForceFlush)
+                {
+                    return;
+                }
+            }
+            else if (this.ParentProvider == null)
+            {
+                return;
+            }
```

Comment:
Could we simplify this a bit?

```csharp
            if (!this.canForceFlush)
            {
                return;
            }

            if (this.tracerProvider == null)
            {
                this.tracerProvider = this.ParentProvider as TracerProvider;
                this.canForceFlush = this.tracerProvider != null;
                if (!this.canForceFlush)
                {
                    return;
                }
            }
```

`ParentProvider` will be set before `OnEnd` is called. Mostly certain about that 😄

Reply:
If we are sure that "`ParentProvider `will be set before `OnEnd` is called" and this behaviour will not be changed then we can simplify the condition. I can mention in the comment that it's assumed `ParentProvider` isn't a null when the `OnEnd` is called. Otherwise for safety we may keep the code as is, because anyway next method execution there will be either `this.canForceFlush = fasle` or `this.tracerProvider == null`, i.e. the condition `this.ParentProvider != null` will not be even checked.

Reply:
Does this need to be checked on every `OnEnd` call? Could this done in the ctor? This check doesn't use the input `Activity` at all.

Reply:
In ctor it's not guaranteed that the `this.ParentProvider` is already set (see this [comment,](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/297#discussion_r851254737) please).